### PR TITLE
Use PRNG in sim/ files

### PR DESCRIPTION
**Pull request by yuzeh**
Edit 3: Okay, we're good to go, pending CI on d98731b.
Edit 2: Hold on, there's a bit more to this commit.
Edit: This PR is ready for review. Should fix #4365.

-----
This PR is not yet complete because some tests are broken, which somewhat confuse me (perhaps because I'm not as familiar with the mechanics of PS/cartridge games).

For example, the Thick Fat test failed with this message:
```
  2) Thick Fat
       should halve damage from Fire- or Ice-type attacks:
     AssertionError [ERR_ASSERTION]: 97 ∈ '[56, 66]'
      at Context.<anonymous> (test/simulator/abilities/thickfat.js:22:10)

  3) Thick Fat
       should be suppressed by Mold Breaker:
     AssertionError [ERR_ASSERTION]: 147 ∈ '[85, 101]'
      at Context.<anonymous> (test/simulator/abilities/thickfat.js:34:10)
```
but succeeded once I fixed the gender of the Thick Fat Pokemon.

The other broken tests (I'll get to investigating these later today):

```

  1) Dancer
       should not copy a move that failed or was blocked by Protect:
     AssertionError [ERR_ASSERTION]: Expected value to remain as undefined, not to change to [object Object].
      at Context.<anonymous> (test/simulator/abilities/dancer.js:88:10)

  2) Emergency Exit
       should prevent Volt Switch after-switches:
     AssertionError [ERR_ASSERTION]: Expected p1: Golisopod species to be Clefable, not Golisopod.
      at Context.<anonymous> (test/simulator/abilities/emergencyexit.js:61:10)

  3) Emergency Exit
       should be suppressed by Sheer Force:

      AssertionError [ERR_ASSERTION]: 'switch' === 'move'
      + expected - actual

      -switch
      +move

      at Context.<anonymous> (test/simulator/abilities/emergencyexit.js:106:10)

  4) Burn
       should halve damage from most Physical attacks:

      AssertionError [ERR_ASSERTION]: Expected p2a: Sableye to be hurt by 41, not by 43.
      + expected - actual

      -43
      +41

      at Context.<anonymous> (test/simulator/misc/statuses.js:32:10)

```

Guidance would be appreciated here!

### CODE REVIEW

**Comment by Zarel** [comment]
Some tests rely on exact RNG behavior; the Burn test specifically looks like you should probably just update the number.

**Comment by Zarel** [comment]
I would honestly recommend just hardcoding the new values, rather than screwing around with `resetRNG`.

**Comment by Zarel** [comment]
Yeah, sorry, when I said "update the number", I mean the number they're expecting. Like with the error "Expected p2a: Sableye to be hurt by 42, not by 43", I meant to update the number to expect Sableye to be hurt by 43. I definitely do not mean to update the seed.

**Comment by Zarel** [comment]
We should also comment `// hardcoded to RNG seed` on these, so we know that it's safe to update them if they break with RNG updates.

File: test/simulator/abilities/dancer.js [hunk 1]
```diff
@@ -54,7 +54,8 @@ describe('Dancer', function () {
 	});
 
 	it('should not copy a move that failed or was blocked by Protect', function () {
-		battle = common.createBattle({gameType: 'doubles', seed: [1, 2, 3, 4]});
+		// hardcoded to RNG seed
```
Comment (Zarel) [inline test/simulator/abilities/dancer.js]: Why are you changing the seed again?

Blah, in general this file needs way more comments.

Comment (yuzeh) [reply test/simulator/abilities/dancer.js]: I can look into this one later, but this is a long sequence of events that seems to be dependent on the RNG.

Comment (Zarel) [reply test/simulator/abilities/dancer.js]: Yeah, this is why I updated the test API to make it easier to tell what the sequence of events actually is... :/

File: test/simulator/abilities/unaware.js [hunk 2]
```diff
@@ -41,8 +41,8 @@ describe('Unaware', function () {
 
 	it('should ignore defense stage changes when Pokemon with it attack', function () {
 		battle = common.createBattle();
-		battle.join('p1', 'Guest 1', 1, [{species: 'Clefable', ability: 'unaware', moves: ['moonblast']}]);
-		battle.join('p2', 'Guest 2', 1, [{species: 'Hariyama', ability: 'thickfat', item: 'laggingtail', moves: ['amnesia']}]);
+		battle.join('p1', 'Guest 1', 1, [{species: 'Clefable', gender: 'M', ability: 'unaware', moves: ['moonblast']}]);
+		battle.join('p2', 'Guest 2', 1, [{species: 'Hariyama', gender: 'M', ability: 'thickfat', item: 'laggingtail', moves: ['amnesia']}]);
```
[the remark this answers is not in the document]
Comment (yuzeh) [reply test/simulator/abilities/unaware.js]: This is a good idea!

File: test/simulator/abilities/dancer.js (hunk 1 above)
Comment (yuzeh) [reply test/simulator/abilities/dancer.js]: In any case, there's no need to change the seed with a call to `resetRNG` after the players join.

File: test/simulator/abilities/thickfat.js [hunk 3]
```diff
@@ -30,6 +31,7 @@ describe('Thick Fat', function () {
 		battle.p2.chooseMove('incinerate').foe.chooseDefault();
 		assert.bounded(target.maxhp - target.hp, [57, 68]);
 		battle.heal(target.maxhp, target, target, battle.getFormat());
+		battle.resetRNG();
 		battle.p2.chooseMove('icebeam').foe.chooseDefault();
 		assert.bounded(target.maxhp - target.hp, [85, 101]);
```
Comment (Zarel) [inline test/simulator/abilities/thickfat.js]: Why is this not bounded in the first place? Is it a crit or something?

I often do `console.log(battle.getDebugLog())` to see what's actually going on.

Comment (yuzeh) [reply test/simulator/abilities/thickfat.js]: As far as I can tell, it was written this way: https://github.com/Zarel/Pokemon-Showdown/commit/7607900160618a598f039722e34ea8e3fa94b968

The original code doesn't rely on any `Math.random` calls, and furthermore the second test was failing when I tried to run it. ¯\_(ツ)_/¯

I'm going to fix the values on this one to the RNG seed.

Comment (yuzeh) [reply test/simulator/abilities/thickfat.js]: Using the default RNG seed without resetting it actually produces a crit in second turn, so we can either change the starting seed or reset the RNG between hits.

Thoughts?

Comment (Zarel) [reply test/simulator/abilities/thickfat.js]: Oh! That explains it. Comment that the crit is intentional, then screw around with seeds or `resetRNG()` to make it crit again.

Comment (yuzeh) [reply test/simulator/abilities/thickfat.js]: Oh wow, so this test is supposed to crit both times instead of crit neither time?

Comment (Zarel) [reply test/simulator/abilities/thickfat.js]: I don't know what the test is _supposed_ to do, but it shouldn't be too hard to figure it out, and then choose a seed to make it do what it's supposed to do...

Comment (yuzeh) [reply test/simulator/abilities/thickfat.js]: Aite, that's done in the latest update. (no crit)

Comment (Zarel) [reply test/simulator/abilities/thickfat.js]: Can you add comments about which ones are supposed to crit and which ones are not supposed to crit?

**Comment by Zarel** [comment]
Looks good.

**Pull request merged**